runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java

```diff
@@ -197,65 +197,85 @@ public int hashCode() {
-     * the collection of entries passed during this function are NOT safe to use
-     * outside the context of this function.
-     *
-     * @param <K>   The type of the key used in projection.
-     * @param <V>   The type of the value used in projection.
-     * @param <I>   The type of the index used in projection.
-     * @param <P>   The type of the projection returned.
+     * Generate a table with a given implementation for the {@link StreamingMap}.
      */
-    @FunctionalInterface
-    public interface ProjectionFunction<K, V, I, P> {
-        @Nonnull
-        Stream<P> generateProjection(I index,
-                                     @Nonnull Stream<Map.Entry<K, V>> entryStream);
-    }
+    public CorfuTable(IndexRegistry<K, V> indices, StreamingMap<K, V> streamingMap,
+                      VersionPolicy versionPolicy) {
+        this.indexSpec = new HashSet<>();
+        this.secondaryIndexes = new HashMap<>();
+        this.mainMap = streamingMap;
+        this.versionPolicy = versionPolicy;
 
-    // The "main" map which contains the primary key-value mappings.
-    private final Map<K,V> mainMap;
-    private final Set<Index<K, V, ? extends Comparable>> indexSpec = new HashSet<>();
-    private final Map<String, Map<Comparable, Map<K, V>>> secondaryIndexes = new HashMap<>();
+        this.optimisticTable = new CorfuTable<>(this.mainMap.getOptimisticMap(), this.indexSpec,
+                this.secondaryIndexes, null);
 
-    /**
-     * Generate a table with a given implementation for the mainMap
-     */
-    public CorfuTable(IndexRegistry<K, V> indices, Map<K, V> mapImpl) {
         indices.forEach(index -> {
             secondaryIndexes.put(index.getName().get(), new HashMap<>());
             indexSpec.add(index);
         });
+
         log.info("CorfuTable: creating CorfuTable with the following indexes: {}",
                 secondaryIndexes.keySet());
-        mainMap = mapImpl;
+    }
+
+    /**
+     * Generate a table with a given implementation for the {@link Map}.
+     */
+    public CorfuTable(IndexRegistry<K, V> indexRegistry, Map<K, V> mapImpl) {
+        this(indexRegistry, new StreamingMapDecorator<>(mapImpl),
+                ICorfuVersionPolicy.DEFAULT);
     }
 
     /**
      * Generate a table with the given set of indexes.
      */
-    public CorfuTable(IndexRegistry<K, V> indices) {
-        this(indices, new HashMap<>());
+    public CorfuTable(IndexRegistry<K, V> indexRegistry) {
+        this(indexRegistry, new StreamingMapDecorator<>(new HashMap<>()));
     }
 
     /**
-     * Default constructor. Generates a table without any secondary indexes.
+     * Generates a table with the given {@link Map} implementation and
+     * without any secondary indexes.
      */
-    public CorfuTable() {
-        this(IndexRegistry.empty(), new HashMap<>());
+    public CorfuTable(Map<K, V> mapImpl) {
+        this(IndexRegistry.empty(), mapImpl);
     }
```

Comment:
Hmm, it seems like this is also changing the current behavior for the in memory CorfuTable with respect to `scanAndFilter` . Now the `scan* ` methods will use more resources, which is not expected.

I think this is another reason to split the types and not overload.

Reply:
Not sure I follow. Can you elaborate?